File: mobile/ios/ShiftPlanner/ShiftPlanner/Features/Manager/Requirements/Repositories/APIRequirementsRepository.swift
import Foundation

final class APIRequirementsRepository: RequirementsRepository {
    private let apiClient: APIClient
    private let calendar = Calendar(identifier: .gregorian)

    init(apiClient: APIClient = .shared) {
        self.apiClient = apiClient
    }

    func fetchPositions() async throws -> [RequirementPositionOption] {
        let request = apiClient.makeRequest(
            path: "positions/",
            method: "GET",
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: [PositionResponseDTO].self)
        return response.map {
            RequirementPositionOption(id: $0.id, name: $0.title)
        }
    }

    func fetchRequirements(startDate: Date, endDate: Date) async throws -> [RequirementOccurrence] {
        let baseURL = apiClient.baseURL.appendingPathComponent("schedule/requirements")
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)
        components?.queryItems = [
            URLQueryItem(name: "start_date", value: dateString(from: startDate)),
            URLQueryItem(name: "end_date", value: dateString(from: endDate))
        ]

        guard let url = components?.url else {
            throw APIClientError.invalidResponse
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let accessToken = apiClient.accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }

        let response = try await apiClient.send(request, as: [ScheduleRequirementResponseDTO].self)
        return try response.map(asOccurrence)
    }

    func createRequirementsBulk(
        startDate: Date,
        endDate: Date,
        weekdays: [Int],
        templates: [RequirementTemplateDraft]
    ) async throws -> [RequirementOccurrence] {
        let payload = ScheduleRequirementBulkCreateDTO(
            startDate: dateString(from: startDate),
            endDate: dateString(from: endDate),
            weekdays: weekdays,
            requirements: templates.map {
                ScheduleRequirementTemplateCreateDTO(
                    positionId: $0.positionId,
                    minStaff: $0.quantity,
                    startTime: timeString(for: $0.startSlot),
                    endTime: timeString(for: $0.endSlot)
                )
            }
        )

        let body = try JSONEncoder().encode(payload)
        let request = apiClient.makeRequest(
            path: "schedule/requirements/bulk",
            method: "POST",
            body: body,
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: ScheduleRequirementBulkResponseDTO.self)
        return try response.requirements.map(asOccurrence)
    }

    func deleteRequirement(id: Int) async throws {
        let request = apiClient.makeRequest(
            path: "schedule/requirements/\(id)",
            method: "DELETE",
            requiresAuthorization: true
        )
        try await apiClient.sendWithoutResponseBody(request)
    }

    private func asOccurrence(_ dto: ScheduleRequirementResponseDTO) throws -> RequirementOccurrence {
        guard let date = Self.dateFormatter.date(from: dto.date) else {
            throw APIClientError.decodingFailed
        }

        let weekday = (calendar.component(.weekday, from: date) + 5) % 7

        return RequirementOccurrence(
            id: dto.id,
            date: date,
            weekday: weekday,
            positionId: dto.positionId,
            positionName: dto.positionTitle,
            quantity: dto.minStaff,
            startSlot: slotIndex(from: dto.startTime),
            endSlot: slotIndex(from: dto.endTime)
        )
    }

    private func dateString(from date: Date) -> String {
        Self.dateFormatter.string(from: date)
    }

    private func timeString(for slot: Int) -> String {
        let totalMinutes = slot * 30
        let hour = totalMinutes / 60
        let minutes = totalMinutes % 60
        return String(format: "%02d:%02d:00", hour, minutes)
    }

    private func slotIndex(from timeString: String) -> Int {
        let parts = timeString.split(separator: ":").compactMap { Int($0) }
        guard parts.count >= 2 else { return 0 }
        return ((parts[0] * 60) + parts[1]) / 30
    }

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = .autoupdatingCurrent
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()
}

File: mobile/android/app/src/main/java/com/froggyriia/shiftplanner/data/requirements/ApiRequirementsRepository.kt
package com.froggyriia.shiftplanner.data.requirements

import com.froggyriia.shiftplanner.data.network.ApiClient
import com.froggyriia.shiftplanner.data.network.ScheduleRequirementBulkCreateDto
import com.froggyriia.shiftplanner.data.network.ScheduleRequirementCreateDto
import com.froggyriia.shiftplanner.data.network.ScheduleRequirementResponseDto
import com.froggyriia.shiftplanner.data.network.ScheduleRequirementTemplateCreateDto
import com.froggyriia.shiftplanner.data.network.ScheduleRequirementUpdateDto
import com.froggyriia.shiftplanner.domain.model.RequirementOccurrence
import com.froggyriia.shiftplanner.domain.model.RequirementPositionOption
import com.froggyriia.shiftplanner.domain.model.RequirementTemplateDraft
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

class ApiRequirementsRepository(
    private val apiClient: ApiClient
) : RequirementsRepository {

    override suspend fun fetchPositions(): List<RequirementPositionOption> = wrap {
        apiClient.api.getPositions().map { RequirementPositionOption(id = it.id, name = it.title) }
    }

    override suspend fun fetchRequirements(
        startDate: String,
        endDate: String
    ): List<RequirementOccurrence> = wrap {
        apiClient.api.getRequirements(startDate = startDate, endDate = endDate)
            .map { it.toOccurrence() }
    }

    override suspend fun createRequirement(
        date: String,
        branchId: Int?,
        positionId: Int,
        quantity: Int,
        startSlot: Int,
        endSlot: Int
    ): RequirementOccurrence = wrap {
        apiClient.api.createRequirement(
            ScheduleRequirementCreateDto(
                branchId = branchId,
                positionId = positionId,
                date = date,
                minStaff = quantity,
                requiredCount = quantity,
                startTime = slotToTime(startSlot),
                endTime = slotToTime(endSlot)
            )
        ).toOccurrence()
    }

    override suspend fun updateRequirement(
        id: Int,
        date: String,
        branchId: Int?,
        positionId: Int,
        quantity: Int,
        startSlot: Int,
        endSlot: Int
    ): RequirementOccurrence = wrap {
        apiClient.api.updateRequirement(
            id = id,
            request = ScheduleRequirementUpdateDto(
                branchId = branchId,
                positionId = positionId,
                date = date,
                minStaff = quantity,
                requiredCount = quantity,
                startTime = slotToTime(startSlot),
                endTime = slotToTime(endSlot)
            )
        ).toOccurrence()
    }

    override suspend fun deleteRequirement(id: Int) = wrap {
        apiClient.api.deleteRequirement(id)
        Unit
    }

    override suspend fun createRequirementsBulk(
        startDate: String,
        endDate: String,
        weekdays: List<Int>,
        templates: List<RequirementTemplateDraft>
    ): List<RequirementOccurrence> = wrap {
        apiClient.api.createRequirementsBulk(
            ScheduleRequirementBulkCreateDto(
                startDate = startDate,
                endDate = endDate,
                weekdays = weekdays,
                requirements = templates.map {
                    ScheduleRequirementTemplateCreateDto(
                        positionId = it.positionId,
                        minStaff = it.quantity,
                        startTime = slotToTime(it.startSlot),
                        endTime = slotToTime(it.endSlot)
                    )
                }
            )
        ).requirements.map { it.toOccurrence() }
    }

    private fun ScheduleRequirementResponseDto.toOccurrence(): RequirementOccurrence {
        val parsedDate = dateFormatter.parse(date) ?: Date()
        val cal = Calendar.getInstance().apply { time = parsedDate }
        // Convert Sunday=1..Saturday=7 to Monday=0..Sunday=6
        val weekday = (cal.get(Calendar.DAY_OF_WEEK) + 5) % 7
        return RequirementOccurrence(
            id = id,
            date = parsedDate,
            weekday = weekday,
            branchId = branchId,
            positionId = positionId,
            positionName = positionTitle,
            quantity = minStaff,
            startSlot = timeToSlot(startTime),
            endSlot = timeToSlot(endTime)
        )
    }

    private fun slotToTime(slot: Int): String {
        val totalMinutes = slot * 30
        return String.format(Locale.US, "%02d:%02d:00", totalMinutes / 60, totalMinutes % 60)
    }

    private fun timeToSlot(time: String): Int {
        val parts = time.split(":").mapNotNull { it.toIntOrNull() }
        if (parts.size < 2) return 0
        return (parts[0] * 60 + parts[1]) / 30
    }

    private val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)

    private suspend fun <T> wrap(block: suspend () -> T): T {
        return try {
            block()
        } catch (error: Throwable) {
            throw Exception(apiClient.userMessage(error))
        }
    }
}
